[docs] Docs for ColossalaiStrategy (#15093)
Lightning-AI/lightning: docs/source-pytorch/advanced/model_parallel.rst

----------

.. _colossalai:

**********
Colossalai
**********

> :class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
> With this chunk mechanism, really large models can be trained with restricted computation devices.
> It supprts larger trainable model size and batch size via reducing CUDA memory fragments and CPU memory consumption.
> Also it speeds up this kind of heterogeneous training by fully utilize all kinds of resources.

Reviewer: ```suggestion
:class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
With this chunk mechanism, really large models can be trained with restricted computation devices.
It supports large trainable model size and batch size by reducing CUDA memory fragments and CPU memory consumption.
Also, it speeds up this kind of heterogeneous training by fully utilizing all kinds of resources.
```

----------

.. _Colossalai:

**************************
Colossalai
**************************

> 

Reviewer 2: We could consider adding concise description here as the intro why or when a Lightning user should consider this strategy. Something along the lines of:

> Use when:
> - Training very large, billion parameter models
> - You need to save cost 
> - Train as big of a model you can given a fixed budget of hardware

This is just a dump example of course. But I would consider something like this before going into the technical part of how it works.

> ColossalAI implements ZeRO-DP with chunk-based memory management.

Reviewer: ```suggestion
:class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
```

            # creat your model here
            self.model = BertForSequenceClassification.from_pretrained("bert-base-uncased")

        def configure_optimizers(self):
            # use the specified optimizer
            optimizer = HybridAdam(self.model.parameters(), self.lr)

        ...

    model = MyBert()
    trainer = Trainer(accelerator="gpu", devices=1, precision=16, strategy="colossalai")
    trainer.fit(model)

You can find more examples `here <https://github.com/hpcaitech/ColossalAI-Pytorch-lightning>`_.

.. note::

    *   The only accelerator which ColossalAI supports is ``"gpu"``. But CPU resources will be used when the placement policy is set to "auto" or "cpu".

    *   The only precision which ColossalAI allows is 16 (FP16).

    *   It only supports a single optimizer, which must be ``colossalai.nn.optimizer.CPUAdam`` or ``colossalai.nn.optimizer.
        HybridAdam`` now. You can set ``adamw_mode`` to False to use normal Adam. Noticing that ``HybridAdam`` is highly optimized, it uses fused CUDA kernel and parallel CPU kernel.
        It is recomended to use ``HybridAdam``, since it updates parameters in GPU and CPU both.

>     *   Your model must be created in the function ``LightningModule.configure_sharded_model()``.

Reviewer: ```suggestion
    *   Your model must be created using the :meth:`~pytorch_lightning.core.module.LightningModule.configure_sharded_model` method.
```

    *   ``ColossalaiStrategy`` doesn't support gradient accumulation now.

.. _colossal_placement_policy:

> Placement Policy
> ================================

Reviewer: ```suggestion
Placement Policy
================
```

> For better training efficiency, placement policies can help users fully exploit their GPU-CPU heterogeneous memory space.
> There are three options for the placement policy.
> They are "cpu", "cuda" and "auto" respectively.
> 
> When the placement policy is set to "cpu", all participated parameters will be offloaded into CPU memory immediately at the end of every auto-grad operation.
> In this way, "cpu" placement policy uses the least CUDA memory.
> It is the best choice for users who want to exceptionally enlarge their model size or training batch size.
> 
> When using "cuda" option, all parameters are placed in the CUDA memory, no CPU resources will be used during the training.
> It is for users who get plenty of CUDA memory.
> 
> The third option, "auto", enables Gemini.
> It monitors the consumption of CUDA memory during the warmup phase and collects CUDA memory usage of all auto-grad operations.
> In later training steps, Gemini manages the data transmission between GPU and CPU automatically according to collected CUDA memory usage information. It is the fastest option when CUDA memory is enough.
> 
> Here's an example of changing the placement policy to "cpu".

Reviewer: ```suggestion
Placement policies can help users fully exploit their GPU-CPU heterogeneous memory space for better training efficiency.
There are three options for the placement policy.
They are "cpu", "cuda" and "auto" respectively.

When the placement policy is set to "cpu", all participated parameters will be offloaded into CPU memory immediately at the end of every auto-grad operation.
In this way, "cpu" placement policy uses the least CUDA memory.
It is the best choice for users who want to exceptionally enlarge their model size or training batch size.

When using "cuda" option, all parameters are placed in the CUDA memory, no CPU resources will be used during the training.
It is for users who get plenty of CUDA memory.

The third option, "auto", enables Gemini.
It monitors the consumption of CUDA memory during the warmup phase and collects CUDA memory usage of all auto-grad operations.
In later training steps, Gemini automatically manages the data transmission between GPU and CPU according to collected CUDA memory usage information. It is the fastest option when CUDA memory is enough.

Here's an example of changing the placement policy to "cpu".
```
